### PR TITLE
Method lookup invariants

### PR DESCRIPTION
**Pull request by kddnewton**
Track assumptions for method lookup.

### CODE REVIEW

File: yjit/src/invariants.rs [hunk 1]
```diff
@@ -15,10 +15,25 @@ use std::collections::HashMap;
...
 /// Used to track all of the various block references that contain assumptions
 /// about the state of the virtual machine.
 pub struct Invariants {
-    bops: HashMap<(RedefinitionFlag, ruby_basic_operators), Vec<BlockRef>>
+    /// Tracks block assumptions about the stability of a basic operator on a
+    /// given class.
+    basic_operators: HashMap<(RedefinitionFlag, ruby_basic_operators), Vec<BlockRef>>,
+
+    /// Tracks block assumptions about callable method entry validity.
+    cme_validity: HashMap<*const rb_callable_method_entry_t, Vec<BlockRef>>,
```
Comment (maximecb) [inline yjit/src/invariants.rs:31]: You can only use `rb_callable_method_entry_t` as a key in a hash map if we're sure that the pointer is not movable by the Ruby GC. Is this the case? If so, that should be commented.

File: yjit/src/invariants.rs [hunk 2]
```diff
@@ -15,10 +15,25 @@ use std::collections::HashMap;
...
+
+    /// Tracks block assumptions about method lookup. Maps a class to a table of
+    /// method ID points to a set of blocks. While a block `b` is in the table,
+    /// b->callee_cme == rb_callable_method_entry(klass, mid).
+    method_lookup: HashMap<VALUE, HashMap<ID, Vec<MethodLookupDependency>>>
```
Comment (maximecb) [inline yjit/src/invariants.rs:36]: We might want to use class serials here for the key? Classes are movable by the Ruby GC if I'm not mistaken? @XrXr

File: yjit/src/cruby.rs [hunk 3]
```diff
@@ -290,7 +290,7 @@ pub struct rb_iseq_t {
         core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
 }
 
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
```
Comment (maximecb) [inline yjit/src/cruby.rs:293]: It's not safe to use any `VALUE` as a key in a hash map because of values movable by the Ruby GC.

**Comment by XrXr** [comment]
Re: GC movement, it's fine to use VALUE as keys as long as we take care to mark them and update them like we do in the C code. The C code uses `st_table` which is also a hash table. It rebuilds the table during GC compaction phases when the GC moves objects around.

**Comment by maximecb** [comment]
Ok then I would recommend mentioning in the comments that the values need to be marked/updated.

File: yjit/src/invariants.rs (hunk 2 above)
Comment (noahgibbs) [reply yjit/src/invariants.rs:36]: The intent is certainly to be able to move them long-term, even if they aren't now. We'd love for compaction to work on everything...

**Comment by kddnewton** [comment]
@maximecb Yeah I wasn't sure about the marking or if it was allowed. The reason I ended up using them as hash keys was because they were already being used that way here: https://github.com/Shopify/ruby/blob/master/yjit_iface.c#L225 and here: https://github.com/Shopify/ruby/blob/master/yjit_iface.c#L228. Basically was just transforming `st_table` structs into `HashMap` objects. I can dig into it a little more and try to find out if we're marking things properly. I'm not entirely sure if we are or not.

**Comment by maximecb** [comment]
We will have to look at making and integration in the GC pretty soon but based on what Alan said, it seems fine to merge this now 👍

**Pull request merged**